README.md
# parallel-numpy-rng
[![tests](https://github.com/lgarrison/parallel-numpy-rng/actions/workflows/test.yml/badge.svg)](https://github.com/lgarrison/parallel-numpy-rng/actions/workflows/test.yml) [![PyPI](https://img.shields.io/pypi/v/parallel-numpy-rng)](https://pypi.org/project/parallel-numpy-rng/)

A multi-threaded random number generator backed by Numpy RNG, with parallelism provided by Numba.

## Overview
Uses the "fast-forward" capability of the [PCG family of RNG](https://www.pcg-random.org),
as exposed by the [new-style Numpy RNG API](https://numpy.org/doc/stable/reference/random/index.html),
to generate random numbers in a multi-threaded manner. The key
is that you get the same random numbers regardless of how many threads were used.

Only a two types of random numbers are supported right now: uniform and normal. More
could be added if there is demand, although some kinds, like bounded random ints, are
hard to parallelize in the approach used here.

The uniform randoms are the same as Numpy produces for a given seed, although the
random normals are not.

## Example + Performance
```python
import numpy as np
import parallel_numpy_rng

seed = 123
parallel_rng = parallel_numpy_rng.default_rng(seed)
numpy_rng = np.random.default_rng(seed)

%timeit numpy_rng.random(size=10**9, dtype=np.float32)                           # 2.85 s
%timeit parallel_rng.random(size=10**9, dtype=np.float32, nthread=1)             # 3.34 s
%timeit parallel_rng.random(size=10**9, dtype=np.float32, nthread=128)           # 67.8 ms

%timeit numpy_rng.standard_normal(size=10**8, dtype=np.float32)                  # 1.12 s
%timeit parallel_rng.standard_normal(size=10**8,dtype=np.float32, nthread=1)     # 1.85 s
%timeit parallel_rng.standard_normal(size=10**8, dtype=np.float32, nthread=128)  # 43.5 ms
```

Note that the `parallel_rng` is slower than Numpy when using a single thread, because the parallel implementation requires a slower algorithm in some cases (this is especially noticeable for float64 and normals)

## Installation
The code works and is [reasonably well tested](./test_parallel_numpy_rng.py), so it's probably ready for use.  It can be installed from PyPI:
```console
$ pip install parallel-numpy-rng
```

## Details
Random number generation can be slow, even with modern algorithms like PCG, so it's helpful to
be able to use multiple threads. The easy way to do this is to give each thread a different
seed, but then the RNG sequence will depend on how many threads you used and how you did the
seed offset. It would be nice if the RNG sequence could be the output of a single logical 
sequence (i.e. the stream resulting from a single seed), and the number of threads could
just be an implementation detail.

The key capability to enable this is cheap fast-forwarding of the underlying RNG.  For example,
if we want to generate *N* random numbers with 2 threads, we know the first thread will do *N/2* calls
to the RNG, thus advancing its internal state that many times. Therefore, we would like to start
the second thread's RNG in a state that is already advanced *N/2* times, but without actually making
*N/2* calls to get there.

This is known as fast-forwarding, or jump-ahead. [Numpy's new RNG API](https://numpy.org/doc/stable/reference/random/index.html)
(as of Numpy 1.17) uses the PCG RNG that supports this feature, and Numpy exposes an [`advance()`
method](https://numpy.org/doc/stable/reference/random/bit_generators/generated/numpy.random.PCG64.advance.html#numpy.random.PCG64.advance)
which implements it.  The new API also exposes CFFI bindings to get PCG random values,
so we can implement the core loop, including parallelism, in a Numba-compiled function
that can call the RNG via a low-level function pointer.

An interesting consequence of using fast-forwarding is that each random value must be generated
with a known number of calls to the underlying RNG, so that we know how many steps to advance
the RNG state by. This means we can't use rejection sampling, which makes a variable number of
calls.  Fortunately, inverse-transform sampling can usually substitute, or more specific methods
like Box-Muller. These can be slower than rejection sampling (or whatever Numpy uses) with one
thread, but even just two threads more than makes up for the difference.


CHANGES.rst
Changelog
=========

0.1.2 (2021-12-29)
------------------

- Packaging changes; release on PyPI.

0.1.1 (2021-12-18)
------------------

Fixes
~~~~~
- ``standard_normal`` sometimes produces ``inf`` [#2]

Enhancements
~~~~~~~~~~~~
- Added more test coverage for mixing functions and threads [#3]


0.1 (2021-12-06)
----------------

- Initial release


test_parallel_numpy_rng.py
import os

import numpy as np
import pytest

# TODO: add PCG64DXSM test

# TODO: repeatedly testing N < Nthread isn't exercising anything new
maxthreads = len(os.sched_getaffinity(0))
all_Nthreads = [1,2,3,4,maxthreads] + \
                list(range(5,maxthreads,12)) + \
                list(range(5,maxthreads,11))
all_Nthreads = sorted(list(set(filter(lambda n: n <= maxthreads, all_Nthreads))))

@pytest.fixture(scope='module')
def allN(request):
    '''~1000 values up to 10^5'''
    _rng = np.random.default_rng(123)
    Ntest = [1,2,3,10,100,1000]
    Ntest += list((10**(_rng.random(size=1000)*5)).astype(int))
    Ntest = sorted(list(set(Ntest)))
    return Ntest

@pytest.fixture(scope='module')
def someN(request):
    '''~100 values up to 10^4'''
    _rng = np.random.default_rng(123)
    Ntest = [1,2,3,10,100,1000]
    Ntest += list((10**(_rng.random(size=100)*4)).astype(int))
    Ntest = sorted(list(set(Ntest)))
    return Ntest


@pytest.fixture(scope='module', params=[123,0xDEADBEEF], ids=['seed1','seed2'])
def seed(request):
    return request.param


@pytest.fixture(scope='module', params=all_Nthreads)
def nthread(request):
    return request.param


@pytest.fixture(scope='module', params=[np.float32,np.float64])
def dtype(request):
    return request.param


@pytest.fixture(scope='module', params=['random','standard_normal'])
def funcname(request):
    return request.param


def test_threads(allN, seed, nthread, dtype, funcname):
    '''do different nthreads give the same answer?
    '''
    
    from parallel_numpy_rng import MTGenerator
    
    for N in allN:
        if N < nthread-1:
            # don't repeatedly test N < nthread
            continue
            
        pcg = np.random.PCG64(seed)
        mtg = MTGenerator(pcg)
        func = getattr(mtg,funcname)
        s = func(size=N, nthread=1, dtype=dtype)

        pcg = np.random.PCG64(seed)
        mtg = MTGenerator(pcg)
        func = getattr(mtg,funcname)
        p = func(size=N, nthread=nthread, dtype=dtype)

        # In theory, different numbers of threads will yield bit-wise identical answers
        # But in practice, the last digit changes sometimes. This is probably because
        # different code paths are taken based on alignment.
        # We will use atol because our values are all O(unity)
        if dtype == np.float32:
            assert np.allclose(s, p, atol=1e-7, rtol=0.)
        elif dtype == np.float64:
            assert np.allclose(s, p, atol=1e-15, rtol=0.)
    
    
def test_resume(someN, seed, nthread, dtype, funcname):
    '''Test that generating an array of randoms with one call
    or several give the same answer
    '''
    
    from parallel_numpy_rng import MTGenerator
    
    rng = np.random.default_rng(seed)
    
    for N in someN:
        if N < nthread-1:
            # don't repeatedly test N < nthread
            continue
            
        pcg = np.random.PCG64(seed)
        mtg = MTGenerator(pcg)
        func = getattr(mtg,funcname)
        a = func(size=N, nthread=nthread, dtype=dtype)
        
        pcg = np.random.PCG64(seed)
        mtg = MTGenerator(pcg)
        func = getattr(mtg,funcname)
        
        res = np.empty(N, dtype=dtype)
        i = 0
        while i < N:
            n = rng.integers(low=1,high=N-i+1)
            res[i:i+n] = func(size=n, nthread=nthread, dtype=dtype)
            i += n

        if dtype == np.float32:
            assert np.allclose(a, res, atol=1e-7, rtol=0.)
        elif dtype == np.float64:
            assert np.allclose(a, res, atol=1e-15, rtol=0.)
            

def test_mixing_threads(someN, seed, nthread, dtype):
    '''Test that changing the number of threads mid-stream
    doesn't matter.  Only standard normal holds any interesting
    external state.
    '''
    funcname = 'standard_normal'
    from parallel_numpy_rng import MTGenerator
    
    rng = np.random.default_rng(seed)
    maxthreads = nthread
    del nthread
    
    for N in someN:
        if N < maxthreads-1:
            # don't repeatedly test N < nthread
            continue
            
        pcg = np.random.PCG64(seed)
        mtg = MTGenerator(pcg)
        func = getattr(mtg,funcname)
        a = func(size=N, nthread=maxthreads, dtype=dtype)
        
        pcg = np.random.PCG64(seed)
        mtg = MTGenerator(pcg)
        func = getattr(mtg,funcname)
        
        res = np.empty(N, dtype=dtype)
        i = 0
        tstart = np.linspace(0, N, maxthreads+1, endpoint=True, dtype=int)
        # sweep from 1 to maxthreads
        for t in range(maxthreads):
            n = tstart[t+1]-tstart[t]
            res[i:i+n] = func(size=n, nthread=t+1, dtype=dtype)
            i += n

        if dtype == np.float32:
            assert np.allclose(a, res, atol=1e-7, rtol=0.)
        elif dtype == np.float64:
            assert np.allclose(a, res, atol=1e-15, rtol=0.)
            
            
def test_mixing_func(someN, seed, nthread, dtype):
    '''Test interleaving random and standard_normal works for different N/thread
    '''
    
    from parallel_numpy_rng import MTGenerator
    
    rng = np.random.default_rng(seed)
    
    for N in someN:
        if N < nthread-1:
            # don't repeatedly test N < nthread
            continue
            
        pcg = np.random.PCG64(seed)
        mtg = MTGenerator(pcg)
        coin_seed = rng.integers(2**16)
        coin_rng = np.random.default_rng(coin_seed)
            
        nchunk = max(2,N//100)
        serial = np.empty(N, dtype=dtype)
        i = 0
        tstart = np.linspace(0, N, nchunk+1, endpoint=True, dtype=int)
        for t in range(nchunk):
            n = tstart[t+1]-tstart[t]
            # in each chunk, flip a coin to decide the function
            func = mtg.random if coin_rng.integers(2) else mtg.standard_normal
            serial[i:i+n] = func(size=n, nthread=1, dtype=dtype)
            i += n
            
            
        pcg = np.random.PCG64(seed)
        mtg = MTGenerator(pcg)
        coin_rng = np.random.default_rng(coin_seed)
        
        parallel = np.empty(N, dtype=dtype)
        i = 0
        for t in range(nchunk):
            n = tstart[t+1]-tstart[t]
            func = mtg.random if coin_rng.integers(2) else mtg.standard_normal
            parallel[i:i+n] = func(size=n, nthread=nthread, dtype=dtype)
            i += n
        
        if dtype == np.float32:
            assert np.allclose(serial, parallel, atol=1e-7, rtol=0.)
        elif dtype == np.float64:
            assert np.allclose(serial, parallel, atol=1e-15, rtol=0.)

    
def test_uniform_matches_numpy(someN, seed, nthread, dtype):
    '''Both Numpy and MTGenerator call the PCG uniform double generator, so
    they actually produce identical streams. Floats are almost the same,
    except we have to reimplement the uint32->float part. So it will be
    close but not exact.
    
    This isn't a property we guarantee or really need to preserve, but
    it's a sign that everything is working as expected.
    '''
    from parallel_numpy_rng import MTGenerator
    
    for N in someN:
        if N < maxthreads-1:
            # don't repeatedly test N < nthread
            continue
            
        pcg = np.random.PCG64(seed)
        mtg = MTGenerator(pcg)
        a = mtg.random(size=N, nthread=nthread, dtype=dtype)
        
        rng = np.random.Generator(np.random.PCG64(seed))
        b = rng.random(size=N, dtype=dtype)
        
        if dtype == np.float64:
            assert np.allclose(a, b, atol=1e-15, rtol=0.)
        elif dtype == np.float32:
            assert np.allclose(a, b, atol=1e-7)
        else:
            raise ValueError(dtype)


def test_finite_normals_float32():
    '''
    If the floats fed to Box-Muller can include 0, it will produce infinity.
    We use the interval (0,1] to avoid this.
    
    In theory, we ought to test float64 the same way. But it's hard to find
    a PCG state that produces 53 zeros...
    
    https://github.com/lgarrison/parallel-numpy-rng/issues/1
    '''
    from parallel_numpy_rng import MTGenerator
    pcg = np.random.PCG64(1194)
    mtg = MTGenerator(pcg)
    a = mtg.standard_normal(size=20000, nthread=maxthreads, dtype=np.float32)
    assert np.all(np.isfinite(a))


src/parallel_numpy_rng/utils.py
'''
Utilities to help generate random values from bits
'''

import numpy as np
from numba import njit

# TODO: for very low latency cases, there may be benefit to inlining these

def generate_int_to_float(bitgen):
    '''Return a nested dict of numba functions, keyed by 'zero'/'nonzero' then dtype'''
    
    # initialize the numba functions to convert ints to floats
    _next_uint32_pcg64 = bitgen().ctypes.next_uint32
    _next_uint64_pcg64 = bitgen().ctypes.next_uint64

    @njit(fastmath=True)
    def _next_float_pcg64(state):
        '''Random float in the semi-open interval [0,1)'''
        return np.float32(np.int32(_next_uint32_pcg64(state) >> np.uint32(8)) * (np.float32(1.) / np.float32(16777216.)))

    @njit(fastmath=True)
    def _next_float_pcg64_nonzero(state):
        '''Random float in the semi-open interval (0,1]'''
        return np.float32((np.int32(_next_uint32_pcg64(state) >> np.uint32(8)) + np.int32(1))  * (np.float32(1.) / np.float32(16777216.)))

    @njit(fastmath=True)
    def _next_double_pcg64(state):
        '''Random double in the semi-open interval [0,1)'''
        return np.float64(np.int64(_next_uint64_pcg64(state) >> np.uint64(11)) * (np.float64(1.) / np.float64(9007199254740992.)))

    @njit(fastmath=True)
    def _next_double_pcg64_nonzero(state):
        '''Random double in the semi-open interval (0,1]'''
        return np.float64((np.int64(_next_uint64_pcg64(state) >> np.uint64(11)) + np.int64(1))  * (np.float64(1.) / np.float64(9007199254740992.)))
    
    _next_float = {'zero': {np.float32:_next_float_pcg64, np.float64:_next_double_pcg64},
                   'nonzero': {np.float32:_next_float_pcg64_nonzero, np.float64:_next_double_pcg64_nonzero},
                  }
    return _next_float


src/parallel_numpy_rng/parallel_numpy_rng.py
'''
File: parallel_numpy_rng.py
Author: Lehman Garrison (https://github.com/lgarrison)
Website: https://github.com/lgarrison/parallel-numpy-rng
License: Apache-2.0
'''

import os

import numba
from numba import njit
import numpy as np

from . import utils

__all__ = ['default_rng', 'MTGenerator']

def default_rng(seed):
    '''A convenience function to create a ``MTGenerator`` with the default
    Numpy bit generator'''
    return MTGenerator(np.random.PCG64(seed))


class MTGenerator:
    '''
    A Multi-Threaded random number generator. Generates the same random number stream
    independent of the number of threads. Implements a subset of the
    ``numpy.random.Generator`` API.
    
    Usage
    -----
    ```
    p = np.random.PCG64(123)  # or PCG64DXSM
    mtg = MTGenerator(p)
    r1 = mtg.random(size=16, nthread=2, dtype=np.float32)
    r2 = mtg.standard_normal(size=16, nthread=2, dtype=np.float32)
    ```
    
    Details
    -------
    This generator lets multiple threads sample from the same logical RNG stream by
    utilizing the fast-forward feature of the underling RNG (probably PCG). This
    requires that generating each random float calls the RNG a constant number of
    times---not true for the default Numpy algorithms, which use rejection sampling!
    Therefore, the algorithms here may be slower than their Numpy counterparts, maybe
    by a lot. But with enough threads, MTGenerator might be faster.
    
    Not all numpy.random.Generator methods are implemented. Some kinds of random
    values, like bounded random ints, are hard to generate without rejection sampling.
    
    Even though we are relying on implementation details of the underlying RNG
    (specifically, how many calls to the RNG were made to generate each output value),
    a key aspect of this is that we don't have to guess if we were right: we can
    query the state of the RNG after generating values to see if it matched our
    guess.
    '''
    def __init__(self, bit_generator, nthread=-1):
        self.bitgen = bit_generator
        if nthread <= 0:
            nthread = len(os.sched_getaffinity(0))
        self.nthread = nthread
        
        # each is a dict, keyed on dtype
        self._next_float = _next_float[bit_generator.state['bit_generator']]['zero']
        self._next_float_nonzero = _next_float[bit_generator.state['bit_generator']]['nonzero']
        
        self._cached_normal = None
        
    def random(self, size=None, nthread=None, out=None, verify_rng=True, dtype=np.float64):
        if size == None:
            size = 1
        if nthread == None:
            nthread = self.nthread
        if nthread < 1:
            raise ValueError("nthread must be >= 1")
        if nthread > size:
            nthread = size
        
        starts = np.linspace(0, size, num=nthread+1, endpoint=True, dtype=np.int64)
        vals_per_call = 2 if dtype == np.float32 else 1
        bitgens = []
        states = np.empty(nthread, dtype=int)
        for t in range(nthread):
            bitgens += [self._copy_bitgen()]
            self._advance_bitgen(bitgens[-1], starts[t], vals_per_call)
            states[t] = bitgens[-1].ctypes.state_address
        
        if out is None:
            out = np.empty(size, dtype=dtype)
        next_float = self._next_float[dtype]
        _random(states, starts, out, next_float)
            
        if verify_rng:
            # Did we advance each RNG by the right amount?
            for t in range(nthread):
                _b = self._copy_bitgen()
                self._advance_bitgen(_b, starts[t+1], vals_per_call)
                assert bitgens[t].state['state'] == _b.state['state']
            
        # finally, advance the base RNG
        self._advance_bitgen(self.bitgen, size, vals_per_call)
        
        return out
    
    @staticmethod
    def _advance_bitgen(bitgen, vals, vals_per_call):
        '''Advance the underlying generator, possibly "fractionally"
        if the generator produces, e.g., two 32-bit values per 64-bit call
        '''
        if bitgen.state['has_uint32']:
            assert vals_per_call == 2  # only supposed to happen with float32
            bitgen.ctypes.next_uint32(bitgen.ctypes.state)
            vals -= 1
        bitgen.advance(vals//vals_per_call)
        for _ in range(vals % vals_per_call):
            bitgen.ctypes.next_uint32(bitgen.ctypes.state)
    
    
    def standard_normal(self, size=None, nthread=None, out=None, verify_rng=True, dtype=np.float64):
        '''
        
        Parameters
        ----------
        verify_rng: bool
            Check the correctness; specifically, that each thread made the number of RNG
            calls expected. Disabling this may improve performance for latency-bound cases.
            Default: True.
        '''
        if size == None:
            size = 1
        if nthread == None:
            nthread = self.nthread
        if nthread > max(size//2,1):
            nthread = max(size//2,1)
        if nthread < 1:
            raise ValueError("nthread must be >= 1")
        if out is None:
            out = np.empty(size, dtype=dtype)
        if size == 0:
            return out
        
        first = 0
        if self._cached_normal != None:
            # the base RNG will already be advanced just past the cached value
            out[0] = self._cached_normal
            self._cached_normal = None
            first = 1  # where to start writing non-cached values
        
        # amount to fast-forward each thread's state
        # force even registration per thread
        # the last thread will spill into a cache if it overflows the array
        ff = 2*np.linspace(0, (size+1-first)//2, num=nthread+1, endpoint=True, dtype=np.int64)
        # note that ff[-1] will be out of bounds because here we're just tracking how much each
        # rng gets advanced, not where it's writing
        
        vals_per_call = 2 if dtype == np.float32 else 1
        bitgens = []
        states = np.empty(nthread, dtype=int)
        for t in range(nthread):
            bitgens += [self._copy_bitgen()]
            bitgens[-1].advance(ff[t]//vals_per_call)
            states[t] = bitgens[-1].ctypes.state_address
        
        # now offset the out array
        _out = out[first:]
        next_float_nonzero = self._next_float_nonzero[dtype]
        _cached_normal = _boxmuller(states, ff, _out, next_float_nonzero)
        if not np.isnan(_cached_normal):
            self._cached_normal = _cached_normal
        del _out
            
        # Did we advance each RNG by the right amount?
        if verify_rng:
            for t in range(nthread):
                _b = self._copy_bitgen()
                _b.advance(ff[t+1]//vals_per_call)
                assert bitgens[t].state['state'] == _b.state['state'], t
            
        # finally, advance the base RNG
        self.bitgen.advance(ff[-1]//vals_per_call)
        
        return out


    @staticmethod
    def _advance_bitgen_boxmuller(bitgen, vals, vals_per_call):
        '''Advance the bitgen for Box-Muller normals
        '''
        bitgen.advance((vals//vals_per_call//2)*2)
        return (vals//vals_per_call) % 2
    
    
    def _copy_bitgen(self):
        '''Return a copy of the base bitgen in its current state'''
        new = self.bitgen.__class__()
        new.state = self.bitgen.state  # this is a deep copy
        return new
    
@njit(fastmath=True, parallel=True)
def _random(states, starts, out, next_double):
    nthread = len(states)
    numba.set_num_threads(max(1,nthread))

    for t in numba.prange(nthread):
        a = starts[t]
        b = starts[t+1]
        s = states[t]
        for i in range(a,b):
            out[i] = next_double(s)
            

@njit(fastmath=True,parallel=True)
def _boxmuller(states, starts, out, next_double):
    nthread = len(states)
    numba.set_num_threads(max(1,nthread))
    dtype = out.dtype.type

    cache = np.full(1, np.nan, dtype=dtype)
    for t in numba.prange(nthread):
        a = starts[t]
        b = min(starts[t+1],len(out))
        s = states[t]
        for i in range(a,b,2):
            u1 = next_double(s)
            u2 = next_double(s)
            amp = np.sqrt(dtype(-2)*np.log(u1))
            ang = dtype(2*np.pi)*u2
            z0 = amp*np.cos(ang)
            z1 = amp*np.sin(ang)

            out[i] = z0
            if i+1 < b:
                out[i+1] = z1
            elif t == nthread-1:
                cache[0] = z1

    return cache[0]

_next_float = {}
_next_float['PCG64'] = utils.generate_int_to_float(np.random.PCG64)

if hasattr(np.random, 'PCG64DXSM'):
    # Numpy >= 1.21
    _next_float['PCG64XDSM'] = utils.generate_int_to_float(np.random.PCG64DXSM)
